### DISCUSSION TITLE
Searching for an object having an attribute with a certain value

### DISCUSSION THREAD

**Discussion by RandomEC** [evennia/evennia#2633] [Evennia Questions] [answered]
As a brief intro, here is the situation. I have two item slots for neck items, finger items and wrist items. For purposes of totaling up skill attribute modifiers on all equipment being worn, it is important to be able to distinguish between these two slots. My solution to this was going to be the following:

 equipment = self.search(equipment, location = caller, attribute_name("equipped") = "neck, first")

"self" in this application is the Character. The problem (which I assume will be apparent to anyone who's done this before ;) ) is in the last search term. What I ideally want to do is search for something with the key stored in the variable "equipment", located on the caller, and having an attribute "equipped", with the value "neck, first".

I've searched (heh) through the documentation and the actual .search function, and I can't figure out how to do the syntax appropriately for this. Or even, actually, if it is possible. As I read the documentation, it seems like you can search for the existence of an attribute on the object, but I can't understand whether you can search for an object having a certain value for that attribute. Is this something that is possible? Is this something that would be better accomplished by storing the equipped location on the object as a tag?

As usual, any assistance is appreciated. I wish I could assure you of the larger volume of questions that I have that I am figuring out the answers to myself ... ;)

Thanks again!

P.S. While I gave Griatch the "answer" on this one, since he squarely answered the question I asked, Tegiminis' answer, which avoided the need for the question entirely also ABSOLUTELY  deserves a read if you are having a similar problem. Thanks to both!

**Comment by Griatch** [comment, marked as answer]
This is not possible with the simple `search` wrapper. You need to make use of a full Django query for this: 

```python
from typeclasses.objects import Object
search_query = "scarf"
equipment = (
    Object.objects
    .get_by_attribute(key="equipment", value="neck, first")
    .filter(db_location=caller, db_key__iexact=search_query)
)
```
If you are not familiar with [Django querie](https://docs.djangoproject.com/en/4.0/topics/db/queries/)s, what this does is to access the `Objects.objects` (which the database manager for all objects that are `Objects`) to use the `.get_by_attribute` helper, which fetches Objects by their Attribute key/value(and category). The return of this (that is, all objects with this combination of Attributes) is then `filter`ed by the location and key you need. They don't have to go on separate lines, it just makes it a little easier to read.

Note that while `.get_by_attribute` is a helper that accepts e.g. `key`, the `.filter` is a standard Django operator that must get the full field name `db_key` etc. 

For marking equipment like this, I'd rather use Tags than Attributes though; put two tags `equipped` and `neck, first` (or three tags `equipped`, `neck` and `first` depending on your structure) on your object and search for them. Only one tag of each type in the entire game and you can combine them in any way. Something like this: 

```python 
equipment = (
    Objects.objects
    .get_by_tag(key="equipment", category="wearables")
    .get_by_tag(key="neck", category="wearables")
    .get_by_tag("first", category="wearables")
    .db_location=caller,
    .db_key="scarf"
)
```

**Comment by RandomEC** [comment]
Thanks for the quick response! A few follow-up questions.

1. To try to use it correctly in the future, is the search wrapper really just intended for things like 'find the object named "scarf" in this room/on this character'? Just less resource-intensive than a full Django query?
2. At the outset of your code, you import Object. I am calling this from inside Character. Doesn't Character inherit from Object? Or am I completely not understanding that concept?
3. So, if I'm understanding the above code, search_query is standing in for whatever the keyword is that I would have passed to the search, right? So, is the Object.objects what is actually calling the Django query?

Thanks again, and I'm loving it so far!

Random

**Comment by Griatch** [comment]
1. The `.search` method Objects is a convenient shortcut for searching for things from commands. It covers 99% of the most common cases (most often you need to find stuff in the same room after all). It can't account for all possibilities though, so for this one needs to dip into the the flexibility of a more direct query of the database. `.search` builds these same queries under the hood, it just abstracts them away for the common case.
2. I assumed here that you wanted to find an 'equipment' which is an `Object` and not a `Character`. The Typeclass you use to base your query on decides the type of thing you search for. So if you did `Character.objects...` you'd only find characters with the given attributes/key etc (if any). 
3. `search_query` (= "scarf" in the example) is the name of the thing you are looking for, while also being a piece of equipment for the neck. The `.objects` property of a Typeclass is a type of handler (a Django manager) that acts as a gateway to the database contents. To make it clearer you can think of `Object.objects.filter()` as _"Let's grab all objects of type `Object` and filter them according to these criteria ..."_. The methods on the `.objects` manager are a mix of Evennia-specific helpers (like `get_by_attribute/tag`) and default Django tools (like `.filter`).

**Reply by RandomEC** [reply]
1. Got it, makes total sense.
2. Nope, you are correct. I misunderstood, and thought that calling .search, on whatever object, searched through all types of objects. I did not realize that it was typeclass-specific. One more thing I've learned today. :)
3. Got it. So, in the bottom of your two examples, you omitted the .filter(). As long as you do db_ references, does the .objects property assume it?

Again, thanks for all the help and the patience. I do so much better with stuff when I learn why things work rather than just learning the thing that works.

**Comment by Griatch** [comment]
2. Actually the `search` helper will always search for Objects, except if you prepend your search query by `*` in which case it will search for Accounts instead. So `caller.search("griatch")` will find my Character whereas `caller.search("*griatch")` will find my Account (in multisession mode 0 they always have the same name). This is an Evennia-ism, only understood by the common search helper. The thing with 'basing' on the class like in `Object.objects.filter(...)` is only relevant for Django querying. 
3. The `db_key*` is the actual database field name. If you were to inspect the database, this is what you'd find. Evennia maps this to `.key` for ease of use normally, so doing `obj.key = foo` will under the hood assign (and save) this to the field `db_key`. When you use the 'raw' Django manager methods though, like `filter`, it does not know anything about the Evennia abstractions and want the raw field names. 

It's worth to read up on this in further details in the docs here (you'll find them by searching for 'search'): https://www.evennia.com/docs/latest/Tutorial-Searching-For-Objects.html

**Reply by RandomEC** [reply]
I had read that a couple of times, but lacked the understanding to appreciate when to use which type of search.  I have a feeling I'm going to have much better comprehension this time through it.

Thanks again.

Random

**Comment by Tegiminis** [comment]
I'm not sure exactly what you're looking to do, but I might propose an alternate idea: it is not actually important to distinguish between any slots of armor. No search needed.

1. Add a dictionary to your character which tracks what you have equipped, rather than having "equipped" be a db value on an item. If you have slots that give stats and slots that don't, separate them out into different dictionaries.
2. Iterate over that dictionary to gather your stat mods and filter out irrelevant ones
3. Apply the stat mods

No explicit django queries necessary. This is what I do for armor in my game.

That said learning Django is obviously very valuable so you do you!

**Reply by RandomEC** [reply]
Tegiminis -

As a practical matter, what you suggest is literally exactly what I am doing. What this question was about was what to do when you have the following corner case:

worn on head: helmet of Itzaca
worn on neck, first: an oath of vengeance
worn on neck, second: an oath of vengeance
worn on body: a well-loved t-shirt
etc.

In this case, when you have two seemingly-identical items (but potentially different due to player-created enchantments or similar), you need a way to distinguish between the two so that you are not adding the stats for one, twice. I don't yet have the ability for players to enchant armor, or for armor to degrade, etc., but I want to be able to distinguish between two incidences of the same piece of armor. And since I've read that doing so with the dbref is gauche, I've been looking for an alternate way to do it.

**Reply by Tegiminis** [reply]
How would you be adding the stats for one twice? Assuming both of those oaths of vengeance are separate items, simply being in different dictionary keys (in this case, neckfirst and necksecond) is enough to distinguish them for purposes of your modifier calculations.

**Reply by RandomEC** [reply]
The dictionary for equipment, as I have written it, takes the form of wear_location:item key. So, when you search for an item that has the key "oath of vengeance" that is located on the caller, you are going to get one of them. When you search for it a second time, you are going to get the same one, or at least I would assume you would.

I'm certainly happy to be missing something here, and to learn something new. :) Should I be storing the item information differently in the dictionary?

Thanks,

Random

**Reply by Tegiminis** [reply]
Evennia allows you to store a reference to a database object in attributes, such that `dict[key]` will return the object in its entirety, no searching necessary. Simply assign the object to the dictionary value. Like `dict[key] = whatever_object`. Since your two oaths of vengeance are different object instances, assigning them to the dictionary this way will automatically differentiate them for future retrieval.

**Reply by RandomEC** [reply]
So, the meat of my wear command is a function on the object itself. So (using the above example), I would just do 

eq_slots["neck, first"] = self

rather than:

eq_slots["neck, first"] = self.key

**Reply by Tegiminis** [reply]
I believe so, yup! Might need to be `self.obj` depending on how the command is set up.

**Reply by RandomEC** [reply]
WOW! That's interesting, and seems very powerful! I'll certainly give that a try instead of my roundabout method!

Thanks!

Random

**Reply by RandomEC** [reply]
Tried this out last night, worked perfectly. It's also opened up much easier ways to do a bunch of things I wanted to do.

Thanks again!
